packages/plugin-react/src/index.ts

```diff
@@ -0,0 +1,258 @@
+
+  const viteBabel: Plugin = {
+    name: 'vite:babel',
+    enforce: 'pre',
+    async transform(code, id, ssr) {
+      if (/\.[tj]sx?$/.test(id)) {
+        const plugins = [...userPlugins]
+
+        const parserPlugins: typeof userParserPlugins = [
+          ...userParserPlugins,
+          'jsx',
+          'importMeta',
+          // This plugin is applied before esbuild transforms the code,
+          // so we need to enable some stage 3 syntax that is supported in
+          // TypeScript and some environments already.
+          'topLevelAwait',
+          'classProperties',
+          'classPrivateProperties',
+          'classPrivateMethods'
+        ]
+
+        const isTypeScript = /\.tsx?$/.test(id)
+        if (isTypeScript) {
+          // TODO: maybe we need to read tsconfig to determine parser plugins to
+          // enable here, but allowing decorators by default since it's very
+          // commonly used with TS.
+          parserPlugins.push('typescript', 'decorators-legacy')
+        }
+
+        const isNodeModules = id.includes('node_modules')
+        if (!isNodeModules && !isProduction) {
+          plugins.push(
+            await interopDefault(
+              import('@babel/plugin-transform-react-jsx-self')
+            ),
+            await interopDefault(
+              import('@babel/plugin-transform-react-jsx-source')
+            )
+          )
+        }
+
+        let useFastRefresh = false
+        if (!skipFastRefresh && !ssr && !isNodeModules) {
+          // Modules with .js or .ts extension must import React.
+          const isReactModule = id.endsWith('x') || code.includes('react')
+          if (isReactModule && filter(id)) {
+            useFastRefresh = true
+            plugins.push([
+              await import('react-refresh/babel'),
+              { skipEnvCheck: true }
+            ])
+          }
+        }
+
+        let ast: t.File | null | undefined
+        if (id.endsWith('x')) {
+          if (opts.jsxRuntime === 'automatic') {
+            // By reverse-compiling "React.createElement" calls into JSX,
+            // React elements provided by dependencies will also use the
+            // automatic runtime!
+            const [restoredAst, isCommonJS] = isNodeModules
+              ? await restoreJSX(babel, code)
+              : [null, false]
+
+            ast = restoredAst
+
+            plugins.push([
+              await import('@babel/plugin-transform-react-jsx'),
+              { runtime: 'automatic' }
+            ])
+
+            // Avoid inserting `import` statements into CJS modules.
+            if (isCommonJS) {
+              plugins.push(babelImportToRequire)
+            }
+          }
+          // Even if the automatic JSX runtime is not used, we can still
+          // inject the React import for .jsx and .tsx modules.
+          else if (!isNodeModules && !/(^|\n)import React /.test(code)) {
+            code = `import React from 'react'; ` + code
+          }
```

Comment:
When testing locally, having this enabled broke the build. The reason is that we already have the `import React from 'react'; ` in our code, so it duplicates the import and breaks stuff.

There should be an option to only do this if the user wants, as this kinda emulates the automatic JSX runtime, and this feature and changes are really recent for most of the projects to have it available.